python/cusignal/io/reader.py

```diff
@@ -0,0 +1,328 @@
+    """
+    Writes binary array to file.
+
+    Parameters
+    ----------
+    file : str
+        A string of filename to store output.
+    binary : ndarray
+        Binary array to be written to file.
+    buffer : ndarray, optional
+        Pinned memory buffer to use when copying data from GPU.
+    append : bool, optional
+        Append to file if created.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing binary data.
+
+    """
+
+    # Get current stream, default or not.
+    stream = cp.cuda.get_current_stream()
+
+    if buffer is None:
+        buffer = cp.asnumpy(binary)
+    else:
+        binary.get(out=buffer)
+
+    if append is True:
+        mode = "ab"
+    else:
+        mode = "wb"
+
+    with open(file, mode) as f:
+        stream.synchronize()
+        f.write(buffer)
+
+
+def unpack_bin(binary, dtype, endianness="L"):
+    """
+    Unpack binary file. If endianness is big-endian, it my be converted
+    to little endian for NVIDIA GPU compatibility.
+
+    Parameters
+    ----------
+    binary : ndarray
+        The binary array to be unpack.
+    dtype : data-type, optional
+        Any object that can be interpreted as a numpy data type.
+    endianness : {'L', 'B'}, optional
+        Data set byte order
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing unpacked binary data.
+
+    """
+
+    if endianness != "L" and endianness != "B" and endianness != "N":
+        raise ValueError("'endianness' should be 'L' or 'B'")
+
+    out = _unpack(binary, dtype, endianness)
+
+    return out
+
+
+def pack_bin(in1):
+    """
+    Pack binary arrary.
+    Data will be packed with little endian for NVIDIA GPU compatibility.
+
+    Parameters
+    ----------
+    in1 : ndarray
+        The ndarray to be pack at binary.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing packed binary data.
+
+    """
+
+    out = _pack(in1)
+
+    return out
+
+
+def read_sigmf(file, num_samples=0, offset=0):
+    """
+    Read and unpack binary file to GPU memory
+
+    Parameters
+    ----------
+    file : str
+        A string of filename to be read/unpacked to GPU.
+    num_samples : int, optional
+        Number of samples to be loaded to GPU. If set to 0,
+        read in all samples.
+    offset : int, optional
+        May be specified as a non-negative integer offset.
+        It is the number of samples before loading 'num_samples'.
+        'offset' must be a multiple of ALLOCATIONGRANULARITY which
+        is equal to PAGESIZE on Unix systems.
+
+    Returns
+    -------
+    out : ndarray
+        An 1-dimensional array containing unpacked binary data.
+
+    """
+
+    meta_ext = ".sigmf-meta"
+    data_ext = ".sigmf-data"
+
+    with open(file + meta_ext, "r") as f:
+        header = json.loads(f.read())
+
+    dataset_type = header["_metadata"]["global"]["core:datatype"]
```

Comment:
This line worries me a bit. We're using the Oracle SigMF recordings where the dtype is found with:

`header["_metadata"]["global"]["core:datatype"]`, but according to: https://github.com/gnuradio/SigMF/issues/114, the dtype can be found via `md["global"]["dtype"]`. Using the latter on our test data would cause an error.

Reply:
New solution: Search entire header for `core:datatype`, which is per spec. I believe `dtype` is incorrect.

```python
def _extract_values(obj, key):
    """Pull all values of specified key from nested JSON."""
    arr = []

    def extract(obj, arr, key):
        """Recursively search for values of key in JSON tree."""
        if isinstance(obj, dict):
            for k, v in obj.items():
                if isinstance(v, (dict, list)):
                    extract(v, arr, key)
                elif k == key:
                    arr.append(v)
        elif isinstance(obj, list):
            for item in obj:
                extract(item, arr, key)
        return arr

    results = extract(obj, arr, key)
    return results
```